generate unified kernel+initrd images that include the root hash as last step of the build + other improvements

poettering opened the pull request:
This adds support for generating combined kernel+initrd image files as last step of the boot that also contain the root hash if verity is enabled. These kernel+initrd images are unified blobs that can be signed as one, in order to implement a complete SecureBoot chain.

lucaswerkmeister commented on mkosi:
```diff
@@ -1228,6 +1255,58 @@ def patch_root_uuid(args, loopdev, root_hash):
+        return
+
+    if args.distribution != Distribution.fedora:
+        return
```
Perhaps it would be cleaner to put everything below in an `install_unified_kernel_fedora` function and make `install_unified_kernel` delegate to the appropriate function for the distribution? Arch Linux, at least, will probably look quite different (use `mkinitcpio` instead of `dracut` (or `kernel-install`?) and then patch together the executable with `objcopy`).

poettering replied:
hmm, i am a bit conservative on adding abstraction code without actually working on at least two backends... Hence I just hacked this up as it is, and then leave the abstraction work to the one who actually adds archlinux support...

(btw, I think it would be good if Arch's mkinitcpio could natively support a `--uefi` switch or so, the way dracut does. I think think objcopy thing is something that better belongs in the initrd implementation than in mkosi. but anyway, this is something for the arch maintainers to figure out)

fsateler replied:
The debian bootable builds also use dracut, so I suppose this can work as-is on debian too?

poettering replied:
@fsateler could be. happy to merge a patch that extends the check if you can verify that it works correctly!

lucaswerkmeister commented on mkosi:
```diff
@@ -1228,6 +1255,58 @@ def patch_root_uuid(args, loopdev, root_hash):
+                dracut += [ "-i", "/usr/lib/systemd/system/systemd-volatile-root.service", "/usr/lib/systemd/system/systemd-volatile-root.service",
+                            "-i", "/usr/lib/systemd/systemd-volatile-root", "/usr/lib/systemd/systemd-volatile-root",
+                            "-i", "/usr/lib/systemd/systemd-veritysetup", "/usr/lib/systemd/systemd-veritysetup",
+                            "-i", "/usr/lib/systemd/system-generators/systemd-veritysetup-generator", "/usr/lib/systemd/system-generators/systemd-veritysetup-generator" ]
```
You could loop over the four files to avoid specifying the long path twice each time, but meh, this works too.

poettering replied:
well, it's mostly a hack, and I kinda hope this can go away quickly, when dracut gets updated to include these things in the "no-host-only" mode anyway. Hence, I wanted it to look a bit like a hack ;-)

poettering replied:
i changed this now a bit and are now using list repitition (i.e. `*2`) to avoid the double paths. I think this looks a bit nicer now!

(In an ideal world we'd use dracut's `-I` switch, instead of `-i` which would remove the need for us to keep repeating each path twice. Unfortunately, I couldn't get it to work — see https://github.com/dracutdevs/dracut/issues/187)

poettering commented on mkosi:
```diff
@@ -1228,6 +1255,58 @@ def patch_root_uuid(args, loopdev, root_hash):
+        if root_hash is not None:
+            cmdline += " roothash=" + root_hash
+
+            for kver in os.scandir(os.path.join(workspace, "root", "usr/lib/modules")):
```
ouch, indeed!

poettering commented:
I force pushed a new version now, which hopefully addresses all issues raised, with the exception of the unified-kernel generalization bits. Let's leave that for the patch that adds similar support for archlinux...

I also added another patch on top that shows the kernel cmdline to use in the summary output

lucaswerkmeister commented:
Alright, looks good to me now, let’s press the green button.

lucaswerkmeister commented:
By the way, this doesn’t add any new dependencies on the host system, correct? Dracut is called inside the image.

poettering commented:
> By the way, this doesn’t add any new dependencies on the host system, correct? Dracut is called inside the image.

Correct!
